### PR TITLE
Feature/roles management

### PR DESCRIPTION
**Pull request by AlexisNeveu**
Role entity definition and tests

### CODE REVIEW

File: src/services/ConfigurationService.js [hunk 1]
```diff
@@ -22,7 +23,7 @@ export function loadConfiguration() {
       liveQuery: {
         classNames: ['Posts', 'Comments'], // List of classes to support for query subscriptions
       },
-      entities: [User, Group],
+      entities: [User, Group, Role],
```
Comment (Zorin95670) [inline src/services/ConfigurationService.js:26]: missing default `role:admin` creation on startup of server

File: src/init.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  }
+}
+
+/**
+ * Function to be called after the start of the api.
+ * @param {object} configuration - The server configuration.
+ * @param {*} Parse - The Parse object to use SDK functionalities.
+ * @returns {Promise} - Promise with nothing on success otherwise an error.
+ */
+export async function upgrade(configuration, Parse = parse) {
```
Comment (Zorin95670) [inline src/init.js:57]: setudDefaultRoles(configuration)

File: src/init.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+async function initParse(configuration, Parse) {
+  Parse.initialize(configuration.parseServer.appId, 'JAVASCRIPT_KEY', configuration.parseServer.masterKey);
+  Parse.serverURL = configuration.parseServer.serverURL;
+}
+/**
+ * Initialize the admin role.
+ * @param {*} Parse - The Parse object to use SDK functionalities.
+ * @returns {Promise} - Promise with the admin role on success otherwise an error.
+ */
+async function initAdminRole(Parse) {
```
Comment (Zorin95670) [inline src/init.js:17]: initRole(roleName)

File: src/init.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
+import parse from 'parse/node.js';
+
+/**
+ * Initialize Parse in order to use sdk functionalities.
+ * @param {object} configuration - The api configuration.
+ * @param {*} Parse - The Parse object to use SDK functionalities.
+ */
+async function initParse(configuration, Parse) {
```
Comment (Zorin95670) [inline src/init.js:8]: initParse(configuration)

File: src/init.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    query.equalTo('name', 'admin');
+    adminRole = await query.first({ useMasterKey: true });
+  } catch (error) {
+    console.error(`Cannot query roles during backend initalization : ${error.code} ${error.message}`);
+    return Promise.reject();
+  }
+  // If the admin role already exists we have nothing to do here
+  if (adminRole) {
+    console.log('Role "admin" already exists');
+    return adminRole;
```
Comment (Zorin95670) [inline src/init.js:32]: return Promise.reject()

File: src/init.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  const acl = new Parse.ACL();
+  acl.setPublicReadAccess(true);
+  acl.setPublicWriteAccess(false);
+  adminRole = new Parse.Role();
+  adminRole.set('name', 'admin');
+  adminRole.setACL(acl);
+  try {
+    adminRole = await adminRole.save({}, { useMasterKey: true });
+    console.log('admin role created');
+    return adminRole;
```
Comment (Zorin95670) [inline src/init.js:44]: return Promise.resolve()

File: src/init.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
+import parse from 'parse/node.js';
```
Comment (Zorin95670) [inline src/init.js:1]: rework on breakline

File: tests/e2e/features/Role.feature [hunk 8]
```diff
@@ -0,0 +1,59 @@
+Feature: Role management
+
+  Scenario: Set and unset a Role to a User
```
Comment (Zorin95670) [inline tests/e2e/features/Role.feature:3]: Add scenario to verify admin role is created.

File: src/init.js (hunk 3 above)
Comment (AlexisNeveu) [reply src/init.js]: done

File: src/init.js (hunk 5 above)
Comment (AlexisNeveu) [reply src/init.js]: done

File: src/init.js (hunk 6 above)
Comment (AlexisNeveu) [reply src/init.js]: modified : return Promise.resolve()

File: src/init.js (hunk 7 above)
Comment (AlexisNeveu) [reply src/init.js]: done

File: tests/e2e/features/Role.feature (hunk 8 above)
Comment (AlexisNeveu) [reply tests/e2e/features/Role.feature:10]: done

File: src/init.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+  let newRole = new Parse.Role();
+  newRole.set('name', roleName);
+  newRole.setACL(acl);
+
+  try {
+    newRole = await newRole.save({}, { useMasterKey: true });
+    console.log(`new ${roleName} role created`);
+
+    return Promise.resolve();
```
Comment (Zorin95670) [inline src/init.js:54]: after reading your new code version, you can remove this line

File: src/init.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+  } catch (error) {
+    console.error(`Error during admin role initialization : ${error.message}`);
+
+    return Promise.reject();
+  }
+}
+
+/**
+ * Function to be called after the start of the api.
```
Comment (Zorin95670) [inline src/init.js:63]: rework on documentation

File: src/init.js [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+ * @param {object} configuration - The server configuration.
+ * @param {*} Parse - The Parse object to use SDK functionalities.
+ * @returns {Promise} - Promise with nothing on success otherwise an error.
+ */
+export async function setupDefaultRoles(configuration, Parse) {
+  initParse(configuration, Parse);
+  console.log('init parse done');
+
+  await initRole('admin', Parse)
```
Comment (Zorin95670) [inline src/init.js:72]: we can only keep `await initRole(...);`

File: src/init.js [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+  const acl = new Parse.ACL();
+  acl.setPublicReadAccess(true);
+  acl.setPublicWriteAccess(false);
+
+  let newRole = new Parse.Role();
+  newRole.set('name', roleName);
+  newRole.setACL(acl);
+
+  try {
+    newRole = await newRole.save({}, { useMasterKey: true });
```
Comment (Zorin95670) [inline src/init.js:51]: newRole is never used, remove it.

File: src/init.js [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    console.log(`new ${roleName} role created`);
+
+    return true;
```
Comment (Zorin95670) [inline src/init.js:54]: put at the end of function

File: src/init.js [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    query.equalTo('name', roleName);
+
+    roleExists = await query.first({ useMasterKey: true });
+  } catch (error) {
+    console.error(`Cannot query roles during backend initalization : ${error.message}`);
+
+    return false;
+  }
+
+  // If the role already exists we have nothing to do here
```
Comment (justinehell) [inline src/init.js:34]:
```suggestion
  // If the role already exists, we have nothing to do here
```

File: src/init.js [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+/**
+ * Setup every default default roles needed at the initialization of the server.
+ * @param {object} configuration - The server configuration.
+ * @param {*} Parse - The Parse object to use SDK functionalities.
+ * @returns {Promise} - Promise with true on success otherwise promise with false.
+ */
+export async function setupDefaultRoles(configuration, Parse) {
+  initParse(configuration, Parse);
```
Comment (justinehell) [inline src/init.js:69]: no need to `await initParse(...)` ?

File: tests/unit/init.spec.js [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+    const configuration = {
+      parseServer: {
+        appId: 'appId',
+        masterKey: 'masterKey',
+        serverURL: 'URL',
+      },
+    };
+    expect(setupDefaultRoles(configuration, Parse)).resolves.toEqual(true);
+  });
+  it('should return a resolved promise with false if query fails', async () => {
```
Comment (justinehell) [inline tests/unit/init.spec.js:96]: jump line before

File: tests/unit/init.spec.js [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+    const configuration = {
+      parseServer: {
+        appId: 'appId',
+        masterKey: 'masterKey',
+        serverURL: 'URL',
+      },
+    };
+    expect(setupDefaultRoles(configuration, Parse)).resolves.toEqual(false);
+  });
+  it('should resolved promise with false if the save fails', async () => {
```
Comment (justinehell) [inline tests/unit/init.spec.js:129]: jump line before

Comment (justinehell) [inline tests/unit/init.spec.js:129]: `should return a revolved promise...`

File: tests/e2e/support/step_definitions/request.js [hunk 18]
```diff
@@ -7,14 +7,16 @@ import nunjucks from 'nunjucks';
  * @param {object} body - Body to pass via HTTP request
  * @param {string} method - HTTP method to use.
  * @param {object} headers - Headers for HTTP request
+ * @param {boolean} form - Whether to convert the body values to URL encoded content.
```
Comment (justinehell) [inline tests/e2e/support/step_definitions/request.js:10]: weird naming `form` for boolean

**Review by justinehell — CHANGES_REQUESTED** [review]

File: src/init.js (hunk 9 above)
Comment (AlexisNeveu) [reply src/init.js]: done

File: src/init.js (hunk 10 above)
Comment (AlexisNeveu) [reply src/init.js]: done

File: src/init.js (hunk 11 above)
Comment (AlexisNeveu) [reply src/init.js]: done

File: src/init.js (hunk 13 above)
Comment (AlexisNeveu) [reply src/init.js]: done

File: src/init.js [hunk 19]
```diff
@@ -0,0 +1,73 @@
+/**
+ * Initialize Parse in order to use sdk functionalities.
+ * @param {object} configuration - The api configuration.
+ * @param {*} Parse - The Parse object to use SDK functionalities.
+ */
```
[the remark this answers is not in the document]
Comment (AlexisNeveu) [reply src/init.js]: done

File: src/init.js [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    return true;
+  } catch (error) {
+    console.error(`Error during admin role initialization : ${error.message}`);
+
+    return false;
+  }
+}
+
+/**
+ * Setup every default default roles needed at the initialization of the server.
```
[the remark this answers is not in the document]
Comment (AlexisNeveu) [reply src/init.js]: done

File: src/init.js (hunk 14 above)
Comment (AlexisNeveu) [reply src/init.js]: done

File: tests/unit/init.spec.js (hunk 16 above)
Comment (AlexisNeveu) [reply tests/unit/init.spec.js:97]: done

File: tests/unit/init.spec.js (hunk 17 above)
Comment (AlexisNeveu) [reply tests/unit/init.spec.js]: done

Comment (AlexisNeveu) [reply tests/unit/init.spec.js]: done

File: tests/e2e/support/step_definitions/request.js (hunk 18 above)
Comment (AlexisNeveu) [reply tests/e2e/support/step_definitions/request.js]: done

File: src/init.js [hunk 21]
```diff
@@ -0,0 +1,72 @@
+/**
+ * Initialize Parse in order to use SDK functionalities.
+ * @param {object} configuration - The api configuration.
+ * @param {*} Parse - The Parse object to use SDK functionalities.
```
Comment (Zorin95670) [inline src/init.js:4]: Missing return

Comment (AlexisNeveu) [reply src/init.js:4]: done

**Review by Zorin95670 — APPROVED** [review]

**Review by justinehell — APPROVED** [review]

**Pull request merged**